ratelimits/bucket.go

```diff
@@ -1,52 +1,253 @@
 	id := ip.String()
-	err := validateIdForName(NewRegistrationsPerIPAddress, id)
+	err := validateIdForName(name, id)
 	if err != nil {
-		return BucketId{}, err
+		return bucketId{}, err
 	}
-	return BucketId{
-		limit:     NewRegistrationsPerIPAddress,
-		bucketKey: joinWithColon(NewRegistrationsPerIPAddress.EnumString(), id),
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), id),
 	}, nil
 }
 
-// NewRegistrationsPerIPv6RangeBucketId returns a BucketId for the /48 IPv6
-// range containing the provided IPv6 address.
-func NewRegistrationsPerIPv6RangeBucketId(ip net.IP) (BucketId, error) {
+// newIPv6RangeCIDRBucketId validates and returns a bucketId for limits that use
+// the 'enum:ipv6RangeCIDR' bucket key format.
+func newIPv6RangeCIDRBucketId(name Name, ip net.IP) (bucketId, error) {
 	if ip.To4() != nil {
-		return BucketId{}, fmt.Errorf("invalid IPv6 address, %q must be an IPv6 address", ip.String())
+		return bucketId{}, fmt.Errorf("invalid IPv6 address, %q must be an IPv6 address", ip.String())
 	}
 	ipMask := net.CIDRMask(48, 128)
 	ipNet := &net.IPNet{IP: ip.Mask(ipMask), Mask: ipMask}
 	id := ipNet.String()
-	err := validateIdForName(NewRegistrationsPerIPv6Range, id)
+	err := validateIdForName(name, id)
+	if err != nil {
+		return bucketId{}, err
+	}
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), id),
+	}, nil
+}
+
+// newRegIdBucketId validates ands returns a bucketId for limits that use the
+// 'enum:regId' bucket key format.
+func newRegIdBucketId(name Name, regId int64) (bucketId, error) {
+	id := strconv.FormatInt(regId, 10)
+	err := validateIdForName(name, id)
+	if err != nil {
+		return bucketId{}, err
+	}
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), id),
+	}, nil
+}
+
+// newDomainBucketId validates and returns a bucketId for limits that use the
+// 'enum:domain' bucket key format.
+func newDomainBucketId(name Name, orderName string) (bucketId, error) {
+	err := validateIdForName(name, orderName)
+	if err != nil {
+		return bucketId{}, err
+	}
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), orderName),
+	}, nil
+}
+
+// newFQDNSetBucketId validates and returns a bucketId for limits that use the
+// 'enum:fqdnSet' bucket key format.
+func newFQDNSetBucketId(name Name, orderNames []string) (bucketId, error) {
+	id := string(core.HashNames(orderNames))
+	err := validateIdForName(name, id)
 	if err != nil {
-		return BucketId{}, err
+		return bucketId{}, err
 	}
-	return BucketId{
-		limit:     NewRegistrationsPerIPv6Range,
-		bucketKey: joinWithColon(NewRegistrationsPerIPv6Range.EnumString(), id),
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), id),
 	}, nil
 }
+
+// Transaction is used to represent a single rate limit Transaction. It contains
+// the bucketId and cost of the Transaction, and flags indicating how the
```

Comment:
Could you document further how `cost` is meant to be chosen? Under what circumstances would it be greater than 1? Are there rate limits for which we know it's always 1 and can embed that knowledge in this package by not taking a `cost` parameter to the corresponding transaction constructor?

Reply:
I can definitely beef this up with some examples, and I believe I can remove cost from a few of the Transaction constructors as well.